### ISSUE TITLE
feat:  support history manifests for application

### ISSUE THREAD

**Issue by penglongli** [argoproj/argo-cd#16614]
# Summary

Currently supports getting the latest Manifests by argo-cd. Such as:
```
$ argocd-uat app manifests test
```

But some users may need to view historical manifests to trace intermediate changes.

# Motivation

Please give examples of your use case, e.g. when would you use this.

- argo cli
```bash
$ argocd app manifests-history test --id 12

### HistoryID: 12
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test
spec:
  replicas: 3
  selector:
    matchLabels:
      app: test
  template:
    metadata:
      labels:
        app: test
    spec:
      containers:
      - image: nginx:1.16.0
        imagePullPolicy: IfNotPresent
...
```

- argo http
```bash
curl -XGET "https://argocd-server.default.svc/api/v1/applications/test/manifestsHistory?historyID=12"
```

# Proposal

We can support the cli and http interface to save it. With this proposal, we need do:
- Save the history manifests (as kubernetes secret)
- Support gRPC interface to get the rivision histories of application

**Comment by crenshaw-dev** [comment]
I'm not sure that the juice is worth the squeeze. I feel like people would be better off implementing the rendered manifest pattern so that the full deployment history is in git.

**Comment by penglongli** [comment]
> I'm not sure that the juice is worth the squeeze. I feel like people would be better off implementing the rendered manifest pattern so that the full deployment history is in git.

Each History stores a copy of the Manifest, so we can easily compare the differences between two different versions.

If it's just Git's submission history, we can't intuitively find the difference between Manifests. The purpose of this HistoryManifest is to trace back the user's recent Resource modifications, or to compare the modifications between certain two times.

**Comment by penglongli** [comment]
> I'm not sure that the juice is worth the squeeze. I feel like people would be better off implementing the rendered manifest pattern so that the full deployment history is in git.

@crenshaw-dev Each History stores a copy of the Manifest, so we can easily compare the differences between two different versions.

If it's just Git's submission history, we can't intuitively find the difference between Manifests. The purpose of this HistoryManifest is to trace back the user's recent Resource modifications, or to compare the modifications between certain two times.

**Comment by crenshaw-dev** [comment]
> If it's just Git's submission history, we can't intuitively find the difference between Manifests

I'm not sure that's true... can't you just find the two commits you're interested in and use the SCM's compare page?

> to trace back the user's recent Resource modifications

Do you mean changes made directly via the UI, or only changes made via GitOps?

**Comment by penglongli** [comment]
> > If it's just Git's submission history, we can't intuitively find the difference between Manifests
> 
> I'm not sure that's true... can't you just find the two commits you're interested in and use the SCM's compare page?
> 
> > to trace back the user's recent Resource modifications
> 
> Do you mean changes made directly via the UI, or only changes made via GitOps?

Yes, we can visually compare the differences in the Manifest generated by the two submissions on the UI.(Not only the UI, but also the command line can be compared after obtaining the manifest of history.)

This is a UI demo, it compares last history with prev history
![image](https://github.com/argoproj/argo-cd/assets/7887287/bf3e2ab8-458f-4c12-9ec0-2ac0a3fbca88)

**Comment by crenshaw-dev** [comment]
I do like the feature a lot. But I think in terms of maintenance cost, we'd be better off providing better first-class support for the "rendered manifest" pattern. i.e. have deep links in the UI to SCM-based diff pages to show the difference between two synced revisions.

**Comment by penglongli** [comment]
> I do like the feature a lot. But I think in terms of maintenance cost, we'd be better off providing better first-class support for the "rendered manifest" pattern. i.e. have deep links in the UI to SCM-based diff pages to show the difference between two synced revisions.

Good idea, but the user can modify the Parameter to create a new History (without any new commits). So the commit-id of the History may be the same in the last few times.
If we want to see the comparison of History, we have to re-render Manifests based on commit-id + parameter.

**Comment by blazehu** [comment]
> > I do like the feature a lot. But I think in terms of maintenance cost, we'd be better off providing better first-class support for the "rendered manifest" pattern. i.e. have deep links in the UI to SCM-based diff pages to show the difference between two synced revisions.
> 
> Good idea, but the user can modify the Parameter to create a new History (without any new commits). So the commit-id of the History may be the same in the last few times. If we want to see the comparison of History, we have to re-render Manifests based on commit-id + parameter.

@crenshaw-dev @penglongli 
+1 I also have such a scenario. If I use SCM's difference page, I can't see the difference after rendering. Since reposerver template rendering often causes a variable to be modified, many resources can be modified. Using SCM alone cannot meet my needs.

**Comment by blakepettersson** [comment]
> Since reposerver template rendering often causes a variable to be modified, many resources can be modified. 

@blazehu - that's what @crenshaw-dev was referring to by "implementing the rendered manifest pattern" - this implies that all the manifests are pre-rendered (perhaps through CI), so that there's no need for the repo-server to run `helm`, `kustomize` etc - just plain old k8s manifests that are ready to go with just a `kubectl apply`.

This has the nice property of WYSIWYG for the manifests in Git vs the destination cluster(s), which means that you'd be able to get your full history from Git. It also has the added bonus of using less resources on the repo-server, since it doesn't need to expend any resources running `helm template` or similar.

Ideally there would be first-class support in Argo CD for doing this, perhaps by extracting out the relevant parts of `GenerateManifests` in the repo-server so that it could be added as an Argo CD CLI command, as well as having more extensive deep-linking support as @crenshaw-dev mentioned.

**Comment by mchtech** [comment]
related issue: https://github.com/argoproj/argo-cd/issues/2355

**Comment by object88** [comment]
Upfront, I am using ArgoCD `v2.6.4+7be094f`

I would like to use this as well.  In my use-case, I want to be able to notify my (semi-technical) release team of when our app-of-apps has completed its deployment.  The release team is technical enough to work with Github to merge a PR, which we will use to change the version of our app-of-apps, but I do not expect them to crawl through the UI before merging to capture existant state, and then again after to report on what was changed.

My intent is to use the notification service to send a webhook to a server I'm writing, which would then reach back to the Argo-CD server, and get the manifests for the current revision and the previous revision, then report on individual apps that have been added, removed, or had a version change.

Example `go` code:
```
	opts := apiclient.ClientOptions{
		GRPCWeb: true,
		ServerAddr: "[REDACTED]",
		AuthToken:  "[REDACTED]",
	}
	c, err := apiclient.NewClient(&opts)
	if err != nil {
		return nil, fmt.Errorf("failed to get client: %v", err)
	}
	closer, appClient, err := c.NewApplicationClient()
	if err != nil {
		return nil, fmt.Errorf("failed to get application client: %v", err)
	}
	defer closer.Close()

	appQuery := application.ApplicationQuery{
		AppNamespace: &namespace,
		Name:         &name,
	}
	app, err := appClient.Get(ctx, &appQuery)
	if err != nil {
		return nil, fmt.Errorf("failed to get app '%s' in namespace '%s': %v", name, namespace, err)
	}

	histories := app.Status.History
	history := histories[len(app.Status.History)-1]
	query := application.ApplicationManifestQuery{
		AppNamespace: &namespace,
		Name:         &name,
		Revision:     &history.Revisions[0],
		Project:      &app.Spec.Project,
	}
	manifest, err := appClient.GetManifests(ctx, &query)
	if err != nil {
		return nil, fmt.Errorf("failed to get manifests: %v", err)
	}
```
This reports back a 502, which matches the behavior with the CLI:
```
$ argocd app manifests dexcare --auth-token=$ARGOCD_TOKEN --server=$ARGOCD_SERVER --grpc-web --revision [REDACTED] --source git
FATA[0006] rpc error: code = Unknown desc = POST https://[REDACTED]/application.ApplicationService/GetManifests failed with status code 502
```

I am using an authToken for a user declared using the Argo-CD helm chart:
```
configs:
  cm:
    accounts.deployment-reporter: apiKey, login
  rbac:
    policy.csv: |
      p, user:deployment-reporter, applications, get, default/*, allow
```

I would expect that I would get back the rendered manifests, and so the sub-applications in the app-of-apps, from which I can extract the Helm data w/ source, targetRevision.

**Comment by olivergondza** [comment]
Now with Source Hydrator in beta with some late features added, I see the barrier of entry is even lower than before. I prefer spending the effort on making it easier to use, then providing an alternative with fracture of its powers.

@argoproj/argocd-approvers-cli, any objections to close?
